### PR TITLE
Overhaul test integration in build system

### PR DESCRIPTION
**Pull request by mqudsi**
This is my second attempt at overhauling our varied test suites and unifying them under the ctest umbrella. Due to the limitations of the way ctest was hacked on to CMake, `test` remains a separate custom target, but nevertheless utilizes `ctest` under the hood.

The biggest change is that each individual test is run in a clean environment, created just for that test. And "test" here no longer means one of three (low level, pexpect, little check) but rather each test run by those test drivers. (Never again will you have to account for an alias that was added but not removed in a previous test).

This change also made it possible to run the tests in parallel (well, mostly), which lets me run the entire fish test suite, including the overhead of creating a new environment for each test, in 13 seconds. Tests continue to be run sequentially under CI.

An additional improvement is the addition of top-level targets for each test, meaning if test `foo` failed and you would like to retry just that test, `ninja -C build test_foo` will retry that test (regardless of which test suite it belongs to).

That said, the ability to run tests in parallel has shown some tests that are designed with global system state in mind - they need to be changed in order to pass when executed in parallel (or else can be forced to execute serially, but I think it's a good idea to clean up the test suite to make everything less brittle) - this applies mainly to `jobs.fish`. ~~The history merge test suite has issues that I believe might indicate a bug in either the creation of a new environment or what constitutes a unique fish session, the history race condition test continues to be extremely flaky and reports items dropped from history when the system is under heavy load when the test is executed - at this point, I suspect an actual issue.~~ Edit: later commits have changed the serialization level for history tests to prevent multiple history tests being run (serially) in one process (inadvertently in the case of this fork, by design in `master`); this seems to have resolved their flakiness.

The actual changes are fairly free of any hacks with one notable exception: cmake/ctest doesn't have a "discover tests" stage so as part of the "configure" step, ~~a bastardized version of `fish_tests` is built (without any dependencies on fish itself) and invoked to emit a list of tests - the alternative to this is for each low level test to be explicitly named in the CMake files (ugly) or each low level test to be moved to a separate source file (cringe).~~

### CODE REVIEW

**Comment by ridiculousfish** [comment]
Thanks for tackling this. I'm very very happy to have real isolated tests and also the ability to run individual tests.

@mqudsi can you say a bit about why you chose CTest? I don't know much about it - is it actually good, or just convenient? I see the configure-time test discovery step, I worry about that since it means you have to reconfigure to pick up new tests.

**Comment by mqudsi** [comment]
The primary reason for going with ctest is to avoid any additional dependencies or further complications to our build system (I feel like they're all ugly, so might as well stick with what we have).

CTest has integration for Google Test if we ever want to formally adopt a framework. Looking through the code for that, I have an idea for how we can avoid the configure-time discovery for the low-level tests suite that I might try my hand at (all other tests discovered via globbing do not require a manual reconfigure to be picked up, as the addition of any files matching the glob pattern for the little check and pexpect tests will automatically trigger a CMake run).

On a more positive note, as of last night all the tests are passing. The build driver was fine, but some of the tests had shared state via the `../test/temp` folder that I patched out, and `jobs.fish` was further isolated to prevent asserting anything about the global state of pre-existing or simultaneously created zombies in the session. macOS runs the tests in a different order, which threw me for a loop because I thought it was an OS-specific issue that was causing them to fail when it turned out to be state (or lack thereof) peristing from a previous test.

**Review by krobelus — COMMENTED** [review]
Running individual `littlecheck` and `pexpect` tests is a great improvement.
With `fish_tests` that already works fairly well, so if test discovery causes problems we could get by without that. Though a better integration is surely nice here as well.

A detail: the individual tests are named just by the basename, like `abbr.fish`. I'd prefer something like `test/abbr.fish` so they are all grouped. I noticed this in the tab-completion results of `ninja` that are  sorted alphabetically by fish, which makes sense since `ninja -t targets` does not give a consistent order.

File: tests/checks/jobs.fish
Comment (krobelus) [inline tests/checks/jobs.fish]: `zombies_among_us` had a nice vibe :)

**Comment by krobelus** [comment]
I noticed is that "unsupported" littlecheck tests are listed as "passed" with ctest while littlecheck would print

image

 Of course that's a really minor detail since we only have like 2 tests with `REQUIRES`.

**Comment by krobelus** [comment]
I noticed is that "unsupported" littlecheck tests are listed as "passed" with ctest while littlecheck would print

image

 Of course that's a really minor detail since we only have like 2 tests with `REQUIRES`.

File: cmake/Tests.cmake
Comment (mqudsi) [reply cmake/Tests.cmake:53]: I developed this branch on Linux, compiling with gcc but linking with lldb.  What’s your toolchain?

File: tests/test_driver.sh
Comment (mqudsi) [reply tests/test_driver.sh:121]: > I noticed is that "unsupported" littlecheck tests are listed as "passed" with ctest while littlecheck would print
> 
> image
> 
> Of course that's a really minor detail since we only have like 2 tests with `REQUIRES`.

If little check returns a specific error code, we can program that to be a “skipped” marker and have ctest act accordingly. I’m guessing it’s just returning 0 here.

**Comment by mqudsi** [comment]
> I noticed is that "unsupported" littlecheck tests are listed as "passed" with ctest while littlecheck would print
> 
> image
> 
> Of course that's a really minor detail since we only have like 2 tests with `REQUIRES`.

If little check returns a specific error code, we can program that to be a “skipped” marker and have ctest act accordingly. I’m guessing it’s just returning 0 here.

File: cmake/Tests.cmake
Comment (krobelus) [reply cmake/Tests.cmake:53]: Linux + gcc (same with clang) + GNU ld

**Comment by krobelus** [comment]
> If little check returns a specific error code, we can program that to be a “skipped” marker

Ok, let's use exit code 125 since git bisect uses the same? Anything other than 0 or 1 should work. We can change littlecheck's behavior here since the `REQUIRE` syntax is fairly new.

**Comment by faho** [comment]
The "wip (do not merge)" label seems superfluous, that's what PR drafts are for - they block merging *in the UI*, so I've taken the liberty of converting this to one.

**Comment by mqudsi** [comment]
> A detail: the individual tests are named just by the basename, like `abbr.fish`. I'd prefer something like `test/abbr.fish` so they are all grouped. I noticed this in the tab-completion results of `ninja` that are sorted alphabetically by fish, which makes sense since `ninja -t targets` does not give a consistent order.

Unfortunately, CMake target names may not contain a slash:

```
CMake Error at cmake/Tests.cmake:101 (add_custom_target):
  add_custom_target called with invalid target name "tests/wait.py".  Target
  names may not contain a slash.  Use ADD_CUSTOM_COMMAND to generate files.
Call Stack (most recent call first):
  cmake/Tests.cmake:155 (add_test_target)
  CMakeLists.txt:207 (include)
```

**Comment by krobelus** [comment]
> Unfortunately, CMake target names may not contain a slash:

Okay, then maybe `test-wait.py` as the next best thing?

**Comment by ridiculousfish** [comment]
@mqudsi do you plan to pick this up again near term? If not I'd be eager to take it over, I'm very excited for this improvement.

**Comment by mqudsi** [comment]
@ridiculousfish, thanks for the ping. I needed some time away from things and when I started dipping my toes back in the water I was too worried about how long I'd abandoned this and how out-of-date it must be by now, which wasn't exactly the kind of motivation I needed. Your enthusiasm, however, is quite inspire. Let me blow the dust off this and see what the deal is; it was already very close to ready iirc, so hopefully it won't be that bad.

**Comment by mqudsi** [comment]
In addition to all the rebasing, I think I've addressed all the issues we've previously talked about including dropping the TMPDIR references from the tests and forgoing the dark magic involved in getting dynamic test discovery for low-level tests working. The tests also have a `test_` prefix when invoked via `ninja` or `cmake` directly, as requested by @krobelus, to make grouping/filtering/ignoring in completions (mentally or actually) easier.

File: tests/test_driver.sh
Comment (krobelus) [reply tests/test_driver.sh:58]: > # RUN: env TMPDIR=%tmpdir %fish %s

Yeah, or just run `set -l tmpdir (mktemp -d)` in the those tests, which also uses TMPDIR.

**Comment by krobelus** [comment]
Maybe let's drop the changes to fish_tests.cpp for now.
They are not so important because to fish_tests binary already works well,
and it's pretty fast.
Those changes should be fairly independent of the rest, so dropping them
makes it easier to review this (my preprocessor-macro suggestion feels wrong)

The main part for me is that we can run individual isolated littlecheck
tests.  For that I've been using a script that mimics the setup done by
`test_fishscript`.  Happy to finally get rid of that.

I find this a bit hard to review because some things are added in earlier
commits, and removed in later ones.

I think the changes to tests/checks/tmux-complete.fish are no longer
necessary and can be dropped. (Or maybe they are only necessary in intermediate states?)
We still want to be able to run tests with littlecheck, so the `mktemp` calls are still good, and removal of the temp dirs is also nice.

Anyway, most of that is not  too important.  We can decide on whether to keep or
drop the fish_tests.cpp changes and then clean up post-merge.

File: tests/test_driver.sh
Comment (faho) [inline tests/test_driver.sh:29]: `command -v`, please. `which` is unnecessary cruft and may not be installed itself.

File: tests/test_driver.sh
Comment (faho) [inline tests/test_driver.sh:82]: This executes one grep per line - why not `$(env | grep)` above?

(yes, technically variables can contain newlines, but then you can't use `env` to begin with)

Comment (faho) [reply tests/test_driver.sh:82]: Alternatively we could just hardcode the LC_ variables we know:

LANG
LC_CTYPE
LC_NUMERIC
LC_TIME
LC_COLLATE
LC_MONETARY
LC_MESSAGES
LC_PAPER
LC_NAME
LC_ADDRESS
LC_TELEPHONE
LC_MEASUREMENT
LC_IDENTIFICATION
LC_ALL

I don't think anyone adds anything more to that silly broken system?

File: build_tools/littlecheck.py
Comment (faho) [inline build_tools/littlecheck.py:9]: I think all or most changes to littlecheck are upstream and should go away on a rebase?

File: cmake/Tests.cmake
Comment (faho) [inline cmake/Tests.cmake:20]: Wow.

File: src/fish_tests.cpp
Comment (faho) [inline src/fish_tests.cpp:4332]: Pretty sure this one typically works for me, WSL 1, provided I do it on the linux filesystem. Fails a ton on OpenBSD, tho.

On the windows filesystem the uvar signalling fails?

File: tests/test_driver.sh
Comment (mqudsi) [reply tests/test_driver.sh:58]: Yes, I left tests that create their own temporary (via mktemp) directory as-is. Only tests that assumed it was OK to write to the shared tests directory or that relied on TEMPDIR or similar were modified to use the temporary `$PWD` instead. I was expecting breakage when executing tests in a random dir as `$PWD` but was pleasantly surprised to see no relative references were used and thus broken by that switch.

File: cmake/Tests.cmake
Comment (mqudsi) [reply cmake/Tests.cmake]: The hack has been removed and CMake now uses text preprocessing combined with a no-op macro in `fish_tests.cpp` to detect low-level tests. Complexity is considerably reduced and imho now at an acceptable level. I am strongly in favor of making this new system holistic, and believe that all test atomics should be runnable via a unified harness/interface.

File: tests/test_driver.sh
Comment (mqudsi) [reply tests/test_driver.sh:29]: Sure.

File: build_tools/littlecheck.py
Comment (mqudsi) [reply build_tools/littlecheck.py:9]: Yes, I believe so. I upstreamed all my littlecheck changes before using them in this branch. I might have botched the rebase, I'm not sure.

File: src/fish_tests.cpp
Comment (mqudsi) [reply src/fish_tests.cpp:4332]: On my WSL machine, the fish source code is in a shared NTFS directory but are you quite sure you're actually running the `history_races` test under WSL? This commit was just re-adding the patch that was dropped in the cleanup of `fish_tests.cpp` in an earlier commit in this branch; I actually (silently) patched out `history_races` under WSL (v1, at least) years ago so there's a good chance you're not actually running it under WSL at all.

File: tests/test_driver.sh
Comment (mqudsi) [reply tests/test_driver.sh:82]: Thanks. I remember having a particular stupid moment with that bit of code, for some reason I was doing something incredibly convoluted based off of trying to port the previous logic line-by-line from fish and not accounting for the differences between `set` and `env`. The entire thing (including both `grep` invocations) can be executed just once.

File: build_tools/littlecheck.py
Comment (mqudsi) [reply build_tools/littlecheck.py:9]: Oh, it's `master` is on a particularly old version of `littlecheck.py`. The `ctest` branch is on a version in between that and littlecheck's `master`. I'll just merge in the latest.

**Comment by mqudsi** [comment]
@krobelus for reviewing, you can actually just completely ignore the changes in the first commit; I'll try to rebase around them but if it's too messy I'm not going to spend too long on that endeavor. I mentioned in another comment but I feel pretty strongly about making all atomic tests runnable via a unified test harness and interface, especially since the complexity it requires out of the build system has been significantly reduced. If it makes you feel any better about the preprocessor suggestion, that was actually my fallback from the start but in hoping to avoid that minor bit of cruft, I ended up with that crazy workaround that I'm not sorry to see go.

**Comment by mqudsi** [comment]
I've rebased again and tried to make the commit history more linear by dropping commits that introduced changes that were later reverted. 

I sporadically see this CI error:

> Testing file checks/psub.fish ... Failure:
> 
>   The CHECK on line 24 wants:
>     foo
> 
>   but there was no remaining output to match.
> 
>   additional output on stderr:1:3:
>     mktemp: X's may only occur at the end of the template
>     usage: mktemp [-d] [-t] [template]
>     note: mktemp is a test function, see tests/test_functions/mktemp.fish
> 
>   Context:
>      <= nothing to match CHECK 'foo' on line 24
>      <= nothing to match CHECK 'bar' on line 25
>      <= nothing to match CHECK 'baz' on line 26
>      <= nothing to match CHECK 'psub file was deleted' on line 34
>      <= nothing to match CHECK 'psub filename ends with .cc' on line 42
>      <= nothing to match CHECK 'psub filename ends with .cc' on line 50
>      <= nothing to match CHECK 'psub directory was deleted' on line 58
> 
>   when running command:
>     ../test/root/bin/fish checks/psub.fish
> 

but `psub.fish` does not call `mktemp` :shrug:

File: tests/test_driver.sh
Comment (ridiculousfish) [inline tests/test_driver.sh:26]: I think this is assuming that the build directory is directly inside the fish source directory, which is often not the case. You can reproduce by making a separate directory and running `cmake /path/to/fish`; the tests will fail.

Comment (mqudsi) [reply tests/test_driver.sh:26]: Do you know if I have to explicitly pass these as command line arguments or environment variables from CMake on invocation or if there's something already set I can use?

Comment (mqudsi) [reply tests/test_driver.sh:26]: nvm. We are (I am?) already passing it as a command line argument, just not using it in the sh script.

**Comment by mqudsi** [comment]
Aside from the issue of littlecheck catching the stderr too late (or something), there appears to have been a bug in the mktemp wrapper function for tests in that it was too sensitive to the presence of an X in the template string. In particular, it was randomly firing when we were calling `mktemp` with a template that contained a directory (e.g. PWD) that was also created via `mktemp` and happened to contain a sequence of `X` as a result. This is patched in the most recent commit.

File: tests/test_driver.sh
Comment (mqudsi) [reply tests/test_driver.sh:35]: Done.

**Comment by faho** [comment]
>but psub.fish does not call mktemp 🤷

`psub` itself does! https://github.com/fish-shell/fish-shell/blob/60a9dcbf0548e0f3445639b2753a9cbd11e8bba7/share/functions/psub.fish#L28-L41

**Comment by mqudsi** [comment]
Thank you - mystery solved! TMPDIR is randomly created by the test driver and then used as part of the template, and it would randomly trigger the overzealous template protection in the test-specific mktemp (now fixed) when it contained something that looked like XX.

File: src/fish_tests.cpp
Comment (faho) [reply src/fish_tests.cpp:4331]: Ah, okay. That's it - it doesn't run the history_races test. I missed that because it doesn't say it's been skipped. It just doesn't mention it at all.

**Comment by mqudsi** [comment]
@zanchey @faho @krobelus @ridiculousfish any objections to merging this? It's been weighing on me and I am loathe to keep rebasing.

**Comment by faho** [comment]
>any objections to merging this?

As far as I can tell all the objections have been dealt with and the tests succeed (and aren't much flakier than what we already have).

So I see no reason to hold this back any longer. Merge if you want!

**Comment by zanchey** [comment]
:shipit:

**Review by krobelus — APPROVED** [review]
Sorry for the delay - this looks awesome, let's merge.

I was confused for a bit because `ninja test_basic.fish` would take 4 seconds but that's just because I had `BUILD_DOCS=ON` which runs sphinx

File: tests/checks/features-qmark1.fish
Comment (mqudsi) [reply tests/checks/features-qmark1.fish:1]: I've added that back in but I think making it conceptually easier to create new tests without having to worry about these things is one of the greatest things about using the test driver.

File: tests/checks/basic.fish
Comment (mqudsi) [reply tests/checks/basic.fish:45]: Sure. Done.

**Comment by mqudsi** [comment]
🎉

thank you, all.

File: tests/checks/features-qmark1.fish
Comment (krobelus) [reply tests/checks/features-qmark1.fish]: Thanks - yeah, I'm probably being overly paranoid here.
If manual invocations ever cause confusion, we could also explicitly prevent them in future.

**Comment by zanchey** [comment]
There are two bits of fallout in the package builders from this merge, neither of which are showstoppers.

It looks like the overrides from 4c2a0de5de have been moved into the test driver, which means the low-level tests go back to running in the user's home directory. This is a bit of an issue on Debian pbuilders which use a non-existent HOME - see d518b01281dfab as well.

Second, old CMake versions call skipped tests "failed". I can't see that this was changed explicitly - it's broken in 3.7 but working in 3.10. This only affects the package builds on LTS version of Debian, so I'll try and chase it down at some stage but it's not a blocker by any means.

**Comment by zanchey** [comment]
> it's broken in 3.7

[  327s] 
[  327s] The following tests FAILED:
[  327s] 	130 - git.fish (Not Run)
[  327s] 	184 - tmux-bind.fish (Not Run)
[  327s] 	185 - tmux-complete.fish (Not Run)
[  327s] 	186 - tmux-prompt.fish (Not Run)
[  327s] Errors while running CTest

> but working in 3.10

The following tests did not run:
	130 - git.fish (Skipped)
	184 - tmux-bind.fish (Skipped)
	185 - tmux-complete.fish (Skipped)
	186 - tmux-prompt.fish (Skipped)

**Comment by faho** [comment]
>Second, old CMake versions call skipped tests "failed". I can't see that this was changed explicitly - it's broken in 3.7 but working in 3.10

Lovely. The alternative here would be to "pass" these tests on our end, maybe even just if cmake < 3.10. It's not like skipped tests are super visible anyway.

**Comment by zanchey** [comment]
> The overrides have indeed been moved into the test driver _but_ at the same time the low-level tests are now executed via the test driver

No, the low-tests are run directly - the test driver always invokes fish first.

```cmake
foreach(LTEST ${LOW_LEVEL_TESTS})
  add_test(
    NAME ${LTEST}
    COMMAND ${CMAKE_BINARY_DIR}/fish_tests ${LTEST}
    WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
  )
  set_tests_properties(${LTEST} PROPERTIES SKIP_RETURN_CODE ${SKIP_RETURN_CODE})
  add_test_target("${LTEST}")
endforeach(LTEST)
```

> I thought maybe the feature to support skipping a test based on the return code was only introduced post CMake 3.7 [but it seems to have been supported since at least version 3.0](https://cmake.org/cmake/help/v3.0/prop_test/SKIP_RETURN_CODE.html)!

They were counted as failures prior to 3.9 (see https://gitlab.kitware.com/cmake/cmake/-/issues/16822).

> Side note: I can't believing viewing the documentation for a CMake API/feature in the latest documentation doesn't include a "introduced in CMake version xxx" or something

Heh. fish's documentation isn't great in that sense either.

**Comment by mqudsi** [comment]
@zanchey  can you please try the ctest_redux branch from  mqudsi/fish-shell? It should fix the debian problem.

**Comment by zanchey** [comment]
@mqudsi, it works great!

**Comment by zanchey** [comment]
Thanks for merging that branch in 93aaa1851e1c9190e6d06e2c5f0d26e434bbd502. Unfortunately that uncovered a new issue in Ubuntu Xenial, which uses CMake 3.5 (which I didn't test until after the merge). The `test` target generate in 3.5 doesn't depend on `fish_tests`! Anyway I think I'll just hack around that in the package build.

**Comment by mqudsi** [comment]
You're most welcome. I saw your patch - that's certainly one way to work around it!

**Pull request merged**
